### ISSUE TITLE
[no-use-before-define] triggered unnecessarily (in functions definitions)

### ISSUE THREAD

**Issue by barroudjo** [typescript-eslint/typescript-eslint#1057]
**Repro**

enable "plugin:@typescript-eslint/recommended"

```TS
function bar(): void {
  console.log(foo);
}
const foo = { foo: 'foo' };
bar();
```

**Expected Result**
There should be no error.
**Actual Result**
eslint gives:
```
'foo' was used before it was defined.eslint(@typescript-eslint/no-use-before-define)
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.3.3` |
| `@typescript-eslint/parser`        | `2.3.3` |
| `TypeScript`                       | `3.6.3` |
| `ESLint`                           | `6.5.1` |
| `node`                             | `12.10.0` |
| `npm`                              | `6.10.3` |

**Comment by bradzacher** [comment]
This is working as expected.

Imagine, for example, that you switch your last two lines around
```ts
function bar(): void {
  console.log(foo);
}
bar();
const foo = { foo: 'foo' };
```

Now your code will error with `Cannot access 'foo' before initialization`.

Attempting to do the full control-flow analysis to detect that `foo` is in fact initialised before `bar` is first called is very, very difficult, and expensive.
Because of this, the rule is built to err on the side of caution.

**Comment by barroudjo** [comment]
Gotcha. That's why it's not enabled by default in most rules set (like standard). Thanks !
